### PR TITLE
feat(core): allow environment based .env files to be loaded

### LINKED ISSUE #15130: .env.production does not override environment variables in .env file
### Current Behavior

a `.env.production` file does not override environment variables in a `.env` file when serving in a production configuration

### Expected Behavior

a `.env.production` file should override environment variables in a `.env` file when serving in a production configuration 

### GitHub Repo

https://github.com/jarodburchill/nx-env-issue-example

### Steps to Reproduce

Setup (can skip if using provided repo):
1. `npx create-nx-workspace nx-env-issue-example --preset=react-monorepo --appName=myapp --bundler=vite --style=css --nxCloud=false`
2. `cd nx-env-issue-example`
3. `echo 'VITE_MOCK=enabled' > ./apps/myapp/.env`
4. `echo 'VITE_MOCK=disabled' > ./apps/myapp/.env.production`
5. `echo '/// <reference types="vite/client" />' > ./apps/myapp/src/vite-env.d.ts` (I personally think this file should be part of the preset template)
6. `echo 'console.log(import.meta.env.VITE_MOCK)' >> ./apps/myapp/src/main.tsx`

Serve in development configuration:
1. `npx nx serve myapp --configuration development`
2. NOTE: browser console logs `enabled` (expected)

Serve in production configuration:
1. `npx nx serve myapp --configuration production`
2. NOTE: browser console logs `enabled` (unexpected, should log `disabled`)

Workaround:
`mv ./apps/myapp/.env ./apps/myapp/.env.development`
with this workaround, when serving in production the browser will now log `disabled` but this workaround requires you to maintain a `.env.[mode]` file for every configuration and not have a shared `.env` file

### Nx Report

```shell
Node : 18.12.1
   OS   : darwin arm64
   npm  : 8.19.2
   
   nx                      : 15.7.0
   @nrwl/linter            : 15.7.0
   @nrwl/workspace         : 15.7.0
   @nrwl/cli               : 15.7.0
   @nrwl/cypress           : 15.7.0
   @nrwl/devkit            : 15.7.0
   @nrwl/eslint-plugin-nx  : 15.7.0
   @nrwl/js                : 15.7.0
   @nrwl/react             : 15.7.0
   @nrwl/tao               : 15.7.0
   @nrwl/vite              : 15.7.0
   typescript              : 4.8.4
```

### Failure Logs

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by SmoshySmosh**
## Current Behavior
Currently, NX allows you to load `.env` files based on a list of [different configurations](https://nx.dev/recipes/environment-variables/define-environment-variables), but it does not provide a way to override the `local` environment name with a user-defined environment name.

## Expected Behavior
This PR introduces the ability to load `.env` files based on a user-defined `[environment-name]`, with the default being `local`. This enhancement provides a more streamlined approach to managing configurations across all targets.

The `-e` and `--environment` flags have been added to allow users to specify a custom environment name when running NX commands. These flags provide greater flexibility and control over environment configurations, simplifying the management of different environments across all targets in NX projects.

It is important to note that this change does not affect the existing `.env` file [loading order](https://nx.dev/recipes/environment-variables/define-environment-variables). The update merely allows for the replacement of `local` with a custom string, offering flexibility while maintaining compatibility with existing configurations.

The updated list of supported `.env` files is as follows:

```
- apps/my-app/.env.[target-name].[configuration-name]
- apps/my-app/.[target-name].[configuration-name].env
- apps/my-app/.env.[target-name]
- apps/my-app/.[target-name].env
- apps/my-app/.env.[environment-name]
- apps/my-app/.[environment-name].env
- apps/my-app/.env
- .env.[target-name].[configuration-name]
- .[target-name].[configuration-name].env
- .env.[target-name]
- .[target-name].env
- .[environment-name].env
- .env.[environment-name]
- .env
```

## Example

```
# Reads from .env.local or apps/my-app/.env.local
nx build my-app
# Reads from .env.beta or apps/my-app/.env.beta
nx build my-app -e beta
nx build my-app --environment beta
```

## Related Issue(s)

Fixes https://github.com/nrwl/nx/issues/15130

### CODE REVIEW

**Comment by SmoshySmosh** [comment]
One of the CircleCI tests failed due to a missing dependency, can someone from the NX team restart that one test? The option is greyed out for me.

Thanks!

`error An unexpected error occurred: "ENOENT: no such file or directory, open '/home/circleci/.cache/yarn/v6/npm-@esbuild-linux-ppc64-0.16.4-69d56c2a960808bee1c7b9b84a115220ec9ce05c-integrity/node_modules/@esbuild/linux-ppc64/.yarn-metadata.json'".
info If you think this is a bug, please open a bug report with the information provided in "/home/circleci/repo/yarn-error.log".`

**Comment by SmoshySmosh** [comment]
Are there any questions or concerns about this change? I’m hoping we can merge it in so it can make the next release and my team can start using it. 

Let me know if there are any adjustments need to be made. 🙂

**Comment by Bielik20** [comment]
I came here looking for exactly that. What I had in mind was:

apps/my-app/.env.[target-name].[configuration-name]
apps/my-app/.[target-name].[configuration-name].env
**apps/my-app/.env.*.[configuration-name]**
**apps/my-app/.*.[configuration-name].env**
apps/my-app/.env.[target-name]
apps/my-app/.[target-name].env
apps/my-app/.env.local
apps/my-app/.local.env
apps/my-app/.env
.env.[target-name].[configuration-name]
.[target-name].[configuration-name].env
**.env.*.[configuration-name]**
**.*.[configuration-name].env**
.env.[target-name]
.[target-name].env
.local.env
.env.local
.env

The additions to the current implementation are in **bold**.
This way we really only adding a special case `target-name` to match any. And it still takes precedence over `.env.local` etc.
Although I like that the proposed solution removes that special `local` case. To be honest I am confused how it is suppose to be used. [Docs](https://nx.dev/recipes/environment-variables/define-environment-variables) do not explain when exactly does `local` take effect - I mean what are the conditions? In case of default configuration? In case it is not CI?

**Comment by Bielik20** [comment]
Hmm, I understand. Still, I would rather we fixed/changed `configuration` behaviour than add yet another variable to keep track off. Configuration is a very useful and works great with DTE. We can run affected with multiple configurations at once. That is HUGE. My concern is that adding something else (`environment`) it would be just too much. What you describe you would want to `environment` to do is, what I think, the `configuration` was designed to accomplish.

Clearly having an ability to target a specific `configuration` no matter the `target` would be very beneficial, arguably more beneficial than being able to target specific `target`.

I would love some maintainers to share their insight though.

**Comment by SmoshySmosh** [comment]
While enhancing the `configuration` option might have been a preferable approach, I found that it was closely tied to the core logic of Nx, making the required modifications quite complex. As such, a breaking change like this should be considered carefully and ideally undertaken by a member of the core team. In light of these factors, I opted for a non-breaking, fully reverse-compatible alternative.

I would appreciate feedback from the Nx team on this PR, as it has been in progress for a few weeks. This update is essential for improving the management of my project's environment, and I am eager to implement the changes.

**Comment by hike-dev** [comment]
[SmoshySmosh](https://github.com/SmoshySmosh)

I have several environments and multiple targets.
Some of targets chained between each other.

It forces me to support full set of .env files per configuration with the same content. And now it looks as a real problem.

As for me - configuration-first is more proper approach. It acceptable to me to not have target related dotenv file at all than support a tones of dotenv files with the same content just because it supports target-first related resolver.

Please make it happen.

**Comment by jensbodal** [comment]
This looks like an exciting proposal, would be curious to know if there's any thoughts from maintainers on this

**Comment by SmoshySmosh** [comment]
@AgentEnder Would you be able to provide some feedback on this PR? I've been keeping it up to date for about a month now without any contributor feedback and I could really use this feature in my projects.

**Comment by SmoshySmosh** [comment]
@vsavkin Could you kindly spare some time to review this PR and provide any feedback or suggestions for changes?

This feature is integral for efficiently managing my applications, services, and e2e tests across different environments. The ability to specify environments with a simple argument would significantly reduce complexity compared to the current methods, such as using `env-cmd` or creating a new `run-command` target for each environment.

**Comment by FrozenPandaz** [comment]
Hey all,

I see the concern with loading the `.env.[configuration-name]` and `apps/app1/.env.[configuration-name]` files.

Customizing this behavior with an environment variable is in my opinion, too much for Nx to handle.

If your application itself needs to load different environment files differently, I'd suggest you use `dotenv` directly within your application to achieve exactly what you want.

So in conclusion, I think we can add loading the target agnostic environment files but this PR adds too much customization which should not be the responsibility of Nx. The PR to add loading the target agnostic environment variables is #17335.
